rdrf/report/reports/generator.py

```diff
@@ -197,141 +153,218 @@ def export_to_json(self, request):
+            def get_flat_json_path(report_model, report_field, i=None):
+                if not report_field:
+                    return None
+                if report_model == 'patient':
+                    prefix = ''
+                else:
+                    prefix = f'{report_model}_'
+
+                field_json_path = re.sub(r"\s", "", report_field)
+                field_json_path = re.sub(r"(.+)({(.*)})", r"\1_\3", field_json_path)
+                if i is not None:
+                    return f"{prefix}{i}_{field_json_path}"
+                else:
+                    return f"{prefix}{field_json_path}"
+
+            def get_variants(lookup_key):
+                query = GqlQuery().fields([lookup_key]).query('dataSummary',
+                                                              input=self.patient_filters).operation().generate()
+                summary_result = self.schema.execute(query, context_value=request)
+                return summary_result.data['dataSummary'][lookup_key]
+
+            fieldnames_dict = OrderedDict()
+            fieldnames_dict['id'] = 'ID'
+
+            # e.g. {'patientAddress': 2}
+            processed_multifield_models = {}
+
+            for rdf in self.report_design.reportdemographicfield_set.all():
+                model_config = self.report_config[rdf.model]
+
+                if rdf.model == 'patient':
+                    # Get label for simple fields
+                    fieldnames_dict[get_flat_json_path(rdf.model, rdf.field)] = model_config['fields'][rdf.field]
+                else:
+                    if model_config.get('multi_field', False):
+                        if not processed_multifield_models.get(rdf.model):
+                            # Lookup how many variants of this model is relevant to our patient dataset
+                            num_variants = get_variants(model_config['variant_lookup'])
+
+                            # Process all the fields for this model now
+                            model_fields = self.report_design.reportdemographicfield_set.filter(
+                                model=rdf.model).values_list("field", flat=True)
+
+                            for i in range(0, (num_variants or 0)):
+                                for mf in model_fields:
+                                    fieldnames_dict[get_flat_json_path(rdf.model, mf,
+                                                                       i)] = f"{model_config['label']}_{model_config['fields'][mf]}_{i}"
+
+                            # Mark as processed
+                            processed_multifield_models[rdf.model] = num_variants
+
+                    else:
+                        fieldnames_dict[get_flat_json_path(rdf.model,
+                                                           rdf.field)] = f"{model_config['label']}_{model_config['fields'][rdf.field]}"
+
+            return fieldnames_dict
+
+        def get_clinical_data_headers(cde_keys):
```

Comment:
This will probably benefit from being moved into a separate file. This class is getting large, and could end up being another mega-class like we see in other parts of the codebase. Worth trying to split it up and write it functionally?

Reply:
Yes, I agree. I might need some pointers with how to write it functionally given my very OO-heavy background :) 
I'll get back to you.